### PR TITLE
Use different rates for physics, robot, and environment update

### LINKED ISSUE #19: Differentiate the update rate of the components of the simulation
Referring to https://github.com/robotology/gym-ignition/pull/18#issuecomment-485236545, we should add support to define different update rates for the computation layer:

1. Physic update rate: integration rate of the physic engine. Currently hardcoded in [`GymFactory`](https://github.com/robotology/gym-ignition/blob/3108cdf1602bebafb9ab81e0f088e81e8e55db30/plugins/GymFactory/src/GymFactory.cpp#L78-L81).
2. Update rate of the robot controller. Currently the class uses the same update rate of the physics.
3. Update rate of the python (or C++) agent. Up to the user.

cc @traversaro

### PR DESCRIPTION
**Pull request by diegoferigo**
Fixes #19

This PR adds the support to different rates for the different components of a RL-based simulation. Particularly:

- **Agent rate**: the rate of the code that creates that reads the environment state, generates the action, and steps the environment. It is typically python code, but with a gympp environment can be also a pure C++ executable.
- **Physics rate**: The rate associated to the step of the physic engine. Small steps are typically associated to more accurate simulations. This number should be kept as big as possible.
- **Controllers rate**: In order to have a better match between simulated and real robot, the low-level motion controllers should be characterized by similar responses. Controllers should run slower than the physics and not faster than the agent.

User code (the agent) usually executes `step()` on the environment. In simulation, it triggers a single call of the gazebo `Run()` function. Since the simulator runs at a higher rate, the number of physics iterations can be calculated as `physics rate / agent rate` and this number is not exposed to the user. Then, the controller lives an independent life, it runs synchronous with the physic engine (very fast) but the PIDs are triggered by a downsampled execution.

### CODE REVIEW

File: gym_ignition/envs/base/ignition_python_env.py [hunk 1]
```diff
@@ -37,40 +34,47 @@ def __init__(self):
         self._gazebo_wrapper = None
         self._model_sdf = None
         self._world_sdf = None
-        self._iterations = 1
-        self._update_rate = 100.0
+        self._iterations = None
+        self._controller_dt = None
+        self._physics_rate = 1000000000.0
```
Comment (traversaro) [inline gym_ignition/envs/base/ignition_python_env.py]: I am a bit confused by this. I would expect the `physics_rate` to be the inverse of the physics integrator time step, but I don't think we are currently using `1 ns` as our physics timestep.

**Comment by diegoferigo** [comment]
cc @traversaro particularly on https://github.com/robotology/gym-ignition/pull/20/commits/058e8baf0d1fb2c385514239d6ee43f7107c62a5 and https://github.com/robotology/gym-ignition/pull/20/commits/3c0b4eb283d2546cfba4e1fd0f0247222f417dbe

File: gym_ignition/envs/base/ignition_python_env.py (hunk 1 above)
Comment (diegoferigo) [reply gym_ignition/envs/base/ignition_python_env.py]: It's unclear to be honest how to set the integrator time step of the simulator. The exposed APIs of ignition gazebo allow set the update rate of the simulation, and it is not clear if it is associated to the physics integration or if within a single simulator iterations the backend can decide to step the physics more than once.

**Comment by traversaro** [comment]
> Since the simulator runs at a higher rate, the number of physics iterations can be calculated as `controller rate / agent rate` and this number is not exposed to the user.

I am a bit confused by this phrase. I am not 100% sure about what you mean by physics iterations, and why this is related to  `controller rate / agent rate` (I am intentionally trying to be a bit naive to simulate a user reading the docs and this discussions).

**Comment by diegoferigo** [comment]
> Is gym-ignition now building with stock ign-gazebo2 or still a patched version?

I updated yesterday my fork and ported the new ecm components on the current upstream version (`ign-gazebo2`). We still need a patched version because not all joint-related components have been implemented upstream. Considering the direction and the discussion we had with them, I updated our components to match the ones that will be likely added upstream.

> There is any reason for setting the update rate of the agent and of the physics update rate via "rate" (i.e. setting the frequency in hertz) and the controller as its period (i.e. by setting the period in seconds)?

It was done in this way due to the exposed APIs. I don't have any strong opinion on it, it is more intuitive to me thinking about the rates in their current form but it's for sure a personal taste.

> For the future it may be useful to call _controller_dt with a less ambiguous name such as joint_controller_dt or something similar, in the end also the agent that is using the gym interface is a controller itself, and ideally we eventually have several nested controllers (for example the walking controllers) running inside the simulation.

I agree on this :+1:

**Comment by traversaro** [comment]
Ok, from what I understand for Gazebo "number of physics iterations" = "number of physics engine time steps"

**Comment by diegoferigo** [comment]
> Ok, from what I understand for Gazebo "number of physics iterations" = "number of physics engine time steps"

It should be like this, but to be honest I don't know if the physics engine can decide to perform smaller steps if required. Otherwise, no variable-steps solvers can be exploited, isn't it?

**Comment by traversaro** [comment]
> It should be like this, but to be honest I don't know if the physics engine can decide to perform smaller steps if required (I mean, within a single gazebo iteration). Otherwise, no variable-steps solvers can be exploited, isn't it?

Good point, even if I think this is not the case now, and in that case the physics system should expose some way of tuning the parameters of the variable-steps solver (such as error tolerance).

**Comment by traversaro** [comment]
> > It should be like this, but to be honest I don't know if the physics engine can decide to perform smaller steps if required (I mean, within a single gazebo iteration). Otherwise, no variable-steps solvers can be exploited, isn't it?
>
> Good point, even if I think this is not the case now, and in that case the physics system should expose some way of tuning the parameters of the variable-steps solver (such as error tolerance). 

The classical name for such "step" is indeed  co-simulation communication step,

File: gym_ignition/envs/base/ignition_python_env.py (hunk 1 above)
Comment (traversaro) [reply gym_ignition/envs/base/ignition_python_env.py]: As far as I could quickly see on the current default branch of `ign-gazebo`, as far as a single `Step` in the ignition-physics plugin correspond to a single step in the underlying fixed-time step integrator, indeed what you have there is  the physics update rate, i.e. the inverse of the physics update time step. For this reason, I am even more curious about this default value of 1 ns for the physics update. Does it make sense? From where does it come from?

**Comment by traversaro** [comment]
> It was done in this way due to the exposed APIs. I don't have any strong opinion on it, it is more intuitive to me thinking about the rates in their current form but it's for sure a personal taste.

Just to understand, it is intuitive for you to think about the agent and physics update using "frequency" (i.e. 1000 Hz), while only for the joint controller think about "timestep" (i.e. 1 ms)? I fail to see the rationale for this, but as long as you are aware of it for me it is fine.

File: plugins/GymFactory/include/gympp/PluginDatabase.h [hunk 2]
```diff
@@ -47,6 +47,9 @@ class GymppPluginRegistrator_CartPole
         cartPoleMetadata.setActionSpaceMetadata(actionSpaceMetadata);
         cartPoleMetadata.setObservationSpaceMetadata(observationSpaceMetadata);
 
+        cartPoleMetadata.setGazeboUpdateRate(1000000000);
```
Comment (traversaro) [inline plugins/GymFactory/include/gympp/PluginDatabase.h]: Again, 1 ns physics time step?

File: plugins/GymFactory/include/gympp/Metadata.h [hunk 3]
```diff
@@ -105,6 +105,9 @@ class gympp::PluginMetadata
     std::string modelFileName;
     std::string worldFileName;
 
+    size_t gazeboUpdateRate;
+    size_t environmentUpdateRate;
```
Comment (traversaro) [inline plugins/GymFactory/include/gympp/Metadata.h]: Any specific reason for only supporting integer rates? 0.5 Hz (i.e. one once every two second) seems to be a perfectly valid  update rate, and also the python interface supports floating point frequencies.

Comment (diegoferigo) [reply plugins/GymFactory/include/gympp/Metadata.h]: Yes these should be `double`

File: gym_ignition/envs/base/ignition_python_env.py (hunk 1 above)
Comment (diegoferigo) [reply gym_ignition/envs/base/ignition_python_env.py]: As discussed f2f the physics rate was descreased to `1000 Hz` by default, and the simulation tries to run as fast as possible setting the RT factor in the sdf file.

File: plugins/GymFactory/include/gympp/PluginDatabase.h (hunk 2 above)
Comment (diegoferigo) [reply plugins/GymFactory/include/gympp/PluginDatabase.h]: Replied https://github.com/robotology/gym-ignition/pull/20#discussion_r281139442

**Comment by diegoferigo** [comment]
For future reference this is how classic gazebo was dealing with physics from sdf: http://gazebosim.org/tutorials?tut=preset_manager&cat=physics.

**Comment by diegoferigo** [comment]
Another interesting link (thanks to @traversaro) to understand the gazebo rates and real-time factor:

http://gazebosim.org/tutorials?tut=physics_params&cat=physics

**Pull request merged**